gobblin-core/src/main/java/org/apache/gobblin/converter/parquet/JsonElementConversionFactory.java

```diff
@@ -0,0 +1,490 @@
+    }
+
+    @Override
+    public Type schema() {
+      String columnName = this.jsonSchema.getColumnName();
+      if (this.repeated) {
+        return Types.repeated(BINARY).as(UTF8).named(columnName);
+      }
+      switch (this.jsonSchema.optionalOrRequired()) {
+        case OPTIONAL:
+          return Types.optional(BINARY).as(UTF8).named(columnName);
+        case REQUIRED:
+          return Types.required(BINARY).as(UTF8).named(columnName);
+        default:
+          throw new RuntimeException("Unsupported Repetition type");
+      }
+    }
+  }
+
+  public static class ArrayConverter extends ComplexConverterForUniformElementTypes {
+    private static final String ARRAY_KEY = "item";
+    private static final String SOURCE_SCHEMA_ITEMS_KEY = "items";
+    private final JsonSchema elementSchema;
+
+    public ArrayConverter(JsonSchema schemaNode, WorkUnitState state) {
+      super(schemaNode, SOURCE_SCHEMA_ITEMS_KEY, true);
+      this.elementSchema = getElementSchema(ARRAY_KEY);
+      super.setElementConverter(getConverter(elementSchema, state, true));
+    }
+
+    @Override
+    Object convertField(JsonElement value) {
+      ParquetGroup array = new ParquetGroup(schema());
+      JsonElementConverter converter = getElementConverter();
+      for (JsonElement elem : (JsonArray) value) {
+        array.add(ARRAY_KEY, converter.convert(elem));
+      }
+      return array;
+    }
+
+    @Override
+    public GroupType schema() {
+      List<Type> fields = new ArrayList<>();
+      fields.add(0, getElementConverter().schema());
+      return new GroupType(this.jsonSchema.optionalOrRequired(), this.jsonSchema.getColumnName(), fields);
+    }
+  }
+
+  public static class EnumConverter extends ComplexConverter {
+    private final JsonSchema elementSchema;
+    private final HashSet<String> symbols = new HashSet<>();
+
+    public EnumConverter(JsonSchema schemaNode, WorkUnitState state) {
+      super(schemaNode, false);
+      elementSchema = getElementSchema();
+      JsonArray symbolsArray = schemaNode.getSymbols();
+      symbolsArray.forEach(e -> symbols.add(e.getAsString()));
+      super.setElementConverter(getConverter(elementSchema, state, false));
+    }
+
+    private JsonSchema getElementSchema() {
+      JsonSchema jsonSchema = JsonSchema.buildBaseSchema(STRING);
+      jsonSchema.setColumnName(this.jsonSchema.getColumnName());
+      return jsonSchema;
+    }
+
+    @Override
+    Object convertField(JsonElement value) {
+      if (symbols.contains(value.getAsString()) || this.jsonSchema.isNullable()) {
+        return this.getElementConverter().convert(value);
+      }
+      throw new RuntimeException("Symbol " + value.getAsString() + " does not belong to set " + symbols.toString());
+    }
+
+    @Override
+    public Type schema() {
+      return this.getElementConverter().schema();
+    }
+  }
+
+  public static class RecordConverter extends ComplexConverter {
```

Comment:
After reading more material about Parquet, I understood its terms and concepts more. 

If I understood your goal correctly, the classes hierarchy will be much clearer if it looks like
```
JsonElementConverter
   |- PrimitiveConverter
   |        |- Int, Double, ...
   |- CollectionConverter
   |        |- Array, Map, Enum
   |- RecordConverter
```

The current class structure enforces some subclass to carry what it shouldn't have from the superclass. Like the `PrimitiveTypeName` argument is hardcoded as `INT32` in complex converter. The `elementConverter` defined in `ComplexConverter` is never used in `RecordConverter`.

With the new classes hierarchy I proposed. It will be at ease for you to move shared logic to its right place without breaking the class semantic.

Reply:
👍